### PR TITLE
fix(pp) fixes a case where pp fails if the pushed content has tags th…

### LINKED ISSUE #29397: PP - pushing content fails with tags that live on another host that does not exist on reciever
### Parent Issue

_No response_

### Problem Statement

If I have a piece of content that lives on system host but uses a tag that lives on another host, one that does not exist on the receiver, the content fails to Push Publish with a totally unhelpful message:

```
Failed to publish because an error occurred: An error occurred when processing Contentlet in '/data/shared/assets/bundles/01J425NKY5MPYKBK2CFKHQ43AB/live/System Host/1/0-content.3ed1f576-57b1-4d19-ab2a-6e5b653ec514.content.xml' with ID '27f4c4d0cdf5c34690da87ce14ffcca3': 'null'
```

### Steps to Reproduce

1. new starter
2. add a new host
3. create a content type with a
- title
- site/folder
- tags
Makes sure it lives on the system host.
4. create a piece of this content on the new host, adding tags, etc.
5. re-save this content on the system host
6. make sure that the tags added live on the new site
7.  try to push the content to demo.dotcms.com

### Acceptance Criteria

If the tag site does not exist, we should save the tag on the system host and not fail the Push publish

### dotCMS Version

*

### Proposed Objective

Quality Assurance

### Proposed Priority

Priority 2 - Important

### External Links... Slack Conversations, Support Tickets, Figma Designs, etc.

_No response_

### Assumptions & Initiation Needs

_No response_

### Quality Assurance Notes & Workarounds

_No response_

### Sub-Tasks & Estimates

_No response_

### PR DESCRIPTION
**Pull request by wezell**
fix(pp) fixes a case where pp fails if the pushed content has tags that live on a host that is not on receiver.

In this case, the tags will instead be saved to the system host.

This PR fixes: #29397

### CODE REVIEW

File: dotCMS/src/enterprise/java/com/dotcms/enterprise/publishing/remote/handler/ContentHandler.java [hunk 1]
```diff
@@ -1049,32 +1051,48 @@ private String getUniqueMatchErrorMsg(final List<Field> uniqueFields, final Stri
...
-						tagAPI.addContentletTagInode(localTag, content.getInode(), fieldVarName);
-					}
+				Host tagSite = Try.of(()->APILocator.getHostAPI().find(remoteTag.getHostId(), APILocator.systemUser(), false)).getOrNull();
+				Host contentSite = Try.of(()->APILocator.getHostAPI().find(content.getIdentifier(), APILocator.systemUser(), false)).getOrNull();
+
+				final String localSiteId = UtilMethods.isSet(()->tagSite.getTagStorage())
+						? tagSite.getTagStorage()
+						: UtilMethods.isSet(()->contentSite.getTagStorage())
+								? contentSite.getTagStorage()
+								: Host.SYSTEM_HOST;
```
Comment (jcastro-dotcms) [inline dotCMS/src/enterprise/java/com/dotcms/enterprise/publishing/remote/handler/ContentHandler.java:1082]: Should we also log a message indicating that the Tag is being saved to System Host? Something similar to what was done when pushing Content Types living in a Site that doesn't exist in the receiver.

Comment (wezell) [reply dotCMS/src/enterprise/java/com/dotcms/enterprise/publishing/remote/handler/ContentHandler.java:1082]: unlike a piece of content, saving a tag to the system host does not really affect permissions and is not a huge deal.  We can, I just did not want to get too noisy.

**Review by jdotcms — APPROVED** [review]

**Review by jcastro-dotcms — APPROVED** [review]

**Review by victoralfaro-dotcms — APPROVED** [review]

File: dotCMS/src/enterprise/java/com/dotcms/enterprise/publishing/remote/handler/ContentHandler.java [hunk 2]
```diff
@@ -1049,32 +1051,48 @@ private String getUniqueMatchErrorMsg(final List<Field> uniqueFields, final Stri
...
+				Tag localTag = tagAPI.getTagByNameAndHost(remoteTag.getTagName(), remoteTag.getHostId());
 
-					TagInode localTagInode = tagAPI.getTagInode(localTag.getTagId(), content.getInode(), fieldVarName);
+				String localUserId = Try.of(()->APILocator.getUserAPI().loadUserById(remoteTag.getUserId()).getUserId()).getOrElse(APILocator.systemUser().getUserId());
 
-					// avoid relating tags twice
-					if(localTagInode==null || !Strings.isNullOrEmpty(localTagInode.getTagId())) {
-						tagAPI.addContentletTagInode(localTag, content.getInode(), fieldVarName);
-					}
+				Host tagSite = Try.of(()->APILocator.getHostAPI().find(remoteTag.getHostId(), APILocator.systemUser(), false)).getOrNull();
```
Comment (dsilvam) [inline dotCMS/src/enterprise/java/com/dotcms/enterprise/publishing/remote/handler/ContentHandler.java:1075]: potencial NPE here if `tagSite is null`?

File: dotCMS/src/enterprise/java/com/dotcms/enterprise/publishing/remote/handler/ContentHandler.java [hunk 3]
```diff
@@ -1049,32 +1051,48 @@ private String getUniqueMatchErrorMsg(final List<Field> uniqueFields, final Stri
...
+				Host contentSite = Try.of(()->APILocator.getHostAPI().find(content.getIdentifier(), APILocator.systemUser(), false)).getOrNull();
+
+				final String localSiteId = UtilMethods.isSet(()->tagSite.getTagStorage())
```
Comment (dsilvam) [inline dotCMS/src/enterprise/java/com/dotcms/enterprise/publishing/remote/handler/ContentHandler.java:1078]: Potencial NPE here if `tagSite` is null?

File: dotcms-integration/src/test/java/com/dotcms/enterprise/publishing/remote/handler/ContentHandlerTest.java [hunk 4]
```diff
@@ -68,4 +86,49 @@ public void Test_TrustedListMatcher() {
         assertFalse(TrustedListMatcher.matches(disallowedClass2));
     }
 
+    @Test
```
Comment (dsilvam) [inline dotcms-integration/src/test/java/com/dotcms/enterprise/publishing/remote/handler/ContentHandlerTest.java:89]: If possible let's include the javadoc with: 
```
 * Method to test: 
 * When: 
 * Should: 
```

E.g., 

```
/**
     * Method to test: {@link ExperimentsAPIImpl#start(long, User)}
     * When: The experiment is started but using {@link ExperimentsAPIImpl#startScheduled(long, User)}
     * Should: Generate a new Running Experiment ID
     */
```

**Review by freddyDOTCMS — APPROVED** [review]

**Pull request merged**